sklearn/cluster/_hierarchical_fast.pyx

```diff
@@ -29,11 +29,11 @@ from numpy.math cimport INFINITY
 ###############################################################################
 # Utilities for computing the ward momentum
 
-def compute_ward_dist(cnp.ndarray[DOUBLE, ndim=1, mode='c'] m_1,
-                      cnp.ndarray[DOUBLE, ndim=2, mode='c'] m_2,
-                      cnp.ndarray[INTP, ndim=1, mode='c'] coord_row,
-                      cnp.ndarray[INTP, ndim=1, mode='c'] coord_col,
-                      cnp.ndarray[DOUBLE, ndim=1, mode='c'] res):
+def compute_ward_dist(DOUBLE[::1] m_1,
+                      DOUBLE[:, ::1] m_2,
+                      INTP[::1] coord_row,
+                      INTP[::1] coord_col,
+                      DOUBLE[::1] res):
```

Comment:
In terms of maintenance, we are trying when changing signature to:
 - `const` qualify non-fused typed memoryview to support read-only arrays or memmapped data
 - format signature similarly to `black` (this makes atomic changes such as addition or removal of parameters less noisy in the future)
 - use concrete types defined in `cnp` instead of using aliases
 - mark function as `nogil` when possible

Here is an example for this function for which all items are applicable:
```suggestion
def compute_ward_dist(
   const cnp.float64_t[::1] m_1,
   const cnp.float64_t[:, ::1] m_2,
   const cnp.intp_t[::1] coord_row,
   const cnp.intp_t[::1] coord_col,
   cnp.float64_t[::1] res,
) nogil:
```

Is it possible to change all the modified signature of